pennylane_forest/qpu.py

```diff
@@ -205,4 +216,51 @@ def expval(self, observable):
                 w = self._eigs[Hkey]["eigval"]
                 return w[0] * p0 + w[1] * p1
 
+        # Multi-qubit observable
+        if len(wires) > 1:
+            # In this case, the observable must be a Tensor product
+            assert isinstance(observable, Tensor), "Multi-qubit observable is not in Tensor format"
+
+            # All observables are rotated to be measured in the Z-basis, so we just need to
+            # check which wires exist in the observable, map them to physical qubits, and measure
+            # the product of PauliZ operators on those qubits
+            pauli_obs = sI()
+            for wire in observable.wires:
+                qubit = wire[0]
+                pauli_obs *= sZ(qubit)
+
+            # Preparation Program
+            prep_prog = Program()
+            for instr in self.program.instructions:
+                if isinstance(instr, Gate):
+                    # split gate and wires -- assumes 1q and 2q gates
+                    tup_gate_wires = instr.out().split(" ")
+                    gate = tup_gate_wires[0]
+                    str_instr = str(gate)
+                    # map wires to qubits
+                    for w in tup_gate_wires[1:]:
+                        str_instr += f" {int(w)}"
+                    prep_prog += Program(str_instr)
+
+            if self.readout_error is not None:
+                prep_prog.define_noisy_readout(
+                    qubit, p00=self.readout_error[0], p11=self.readout_error[1]
+                )
+
+            # Measure out multi-qubit observable
+            tomo_expt = Experiment(settings=[ExperimentSetting(TensorProductState(), pauli_obs)],
+                                   program=prep_prog)
+            grouped_tomo_expt = group_experiments(tomo_expt)
+            meas_obs = list(
+                measure_observables(
+                    self.qc,
+                    grouped_tomo_expt,
+                    active_reset=self.active_reset,
+                    symmetrize_readout=self.symmetrize_readout,
+                    calibrate_readout=self.calibrate_readout,
+                )
+            )
```

Comment:
Something that @antalszava pointed out is that `measure_observables` will implicitly execute the circuit on the QPU. However, `QVMDevice.generate_samples()` has already done one execution, making it redundant. This could be solved by simply overwriting `generate_samples()` here to take into account operator estimation, _or_ perhaps by our eventual goal of merging QVM and QPU :laughing:

Reply:
Not sure I follow this point. Is the idea `QVMDevice.generate_samples()` being called before/while this particular call to `measure_observables`?  (and if that's the case, it's not too clear to me where exactly this might be happening)

Reply:
Yes, one missing piece here is that we have moved to a new API in PennyLane and this affects the plugins as well.

Following up on the previous https://github.com/rigetti/pennylane-forest/pull/38, now ForestDevice inherits from QubitDevice. The PennyLane type of flow of logic involves evaluating a quantum circuit by calling on the execute method of the device (such is QPUDevice).

Each instance of the ForestDevice (now QPUDevice in our case) will currently inherit the same execute method [as seen here](https://github.com/XanaduAI/pennylane/blob/a1f22b10ee968ade10488c6c42a7b153553cf628/pennylane/_qubit_device.py#L123) (which is not redefined in any of the subclasses).

As per the logic in `QubitDevice.execute`, currently [`generate_samples` is called](https://github.com/XanaduAI/pennylane/blob/8cccf8681ccfc81f46a771092fcb4442e0319512/pennylane/_qubit_device.py#L158), as the `QPUDevice` does not support analytic evaluation (it being a hardware). After this, [statistics are extracted](https://github.com/XanaduAI/pennylane/blob/8cccf8681ccfc81f46a771092fcb4442e0319512/pennylane/_qubit_device.py#L161) and therefore eventually we [call on `QPUDevice.expval`](https://github.com/XanaduAI/pennylane/blob/a1f22b10ee968ade10488c6c42a7b153553cf628/pennylane/_qubit_device.py#L249) which usually corresponds to `QubitDevice.expval`.

Now in our case, `QPUDevice` inherited the `QVMDevice.generate_samples` method, which makes the call on the `QuantumComputer.run` method. It will in the `expval` method, however, call on `measure_observables` as well.

Reply:
Thanks for the detailed explanation @antalszava! It seems to me this won't be the last time `generate_samples` and `measure_observables` will conflict ;) , mostly because they're both trying to generate samples!

My gut feeling is also that even if we do merge the QVM and QPU classes, we'll still be faced with the choice of either over-riding `generate_samples` to use `measure_observables`, or port over all the logic of error mitigation over to `generate_samples` (and/or its helper functions). I'm not sure how I can over-ride it here though, would appreciate input on that!

Reply:
> I'm not sure how I can over-ride it here though, would appreciate input on that!

I think it's a good question, and not entirely trivial --- I say we merge this as is, and make a new issue to determine how to avoid this conflict!

Reply:
That sounds good to me. I've created a new issue here: https://github.com/rigetti/pennylane-forest/issues/45.